### ISSUE TITLE
Use TypeScript to automatically generate `shield()`

### ISSUE THREAD

**Issue by brillout** [vikejs/telefunc#23]
Let's support only basic types at first, which I expect to be fairly simple to implement.

Most telefunctions have basic types. (The Telefunc idiomatic way of doing things is to implement many tailored telefunctions instead of only few generic ones, see https://telefunc.com/tour#inversion-of-control. Such tailored telefunctions usually have simple types.)

PR welcome.

**Comment by louwers** [comment]
I'd love to give this a go!

**Comment by brillout** [comment]
👍 That would be quite lovely indeed.

What TypeScript parser could we use? Babel I guess? Other alternatives?

**Comment by louwers** [comment]
Didn't look deeply into this, but how about cutting out the middle man? :wink:  

https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API

**Comment by brillout** [comment]
Why not :-).

**Comment by brillout** [comment]
For now, I'd suggest to only support inline types:

```ts
// *.telefunc.js

// Supported
export async function hello(name: string) {
}

// Not supported
type Name = string
export async function hello(name: Name) {
}
````

This will work for most telefunctions. Because telefunctions should be tailored to user events and are, therefore, expected to have simple arguments.

For more advanced support we can have a look at https://www.typescriptneedstypes.com/.

But I'd suggest to first focus on inline types. We can then see how/if we want to extend support.

**Comment by bartenra** [comment]
This looks interesting... https://github.com/typescript-rtti/typescript-rtti#checking-the-type-of-a-value-at-runtime

**Comment by brillout** [comment]
👍 Looks interesting indeed.

I wonder how stable it is though.

 The readme says:

> Please file a bug if you cannot build your codebase with the transformer enabled

Does this mean that it can break apps?

Also, it requires `ttypescript`; I wonder about the potential downsides of using `ttypescript`. Does it only expose TypeScript internals or does it also do other things?

**Comment by louwers** [comment]
So I looked into it a bit, and `ttypescript` patches `typescript` so that it reads which transformers to apply from `tsconfig.json`. Applying transformers is possible using the [public TypeScript API](https://github.com/Microsoft/TypeScript/pull/13940), in fact a transformer is just:

```ts
type Transformer<T extends Node> = (node: T) => T;
```

But you can only use that if you use the Compiler API I mentioned earlier, not if you use `tsc`. All in all I think `ttypescript` does not do a lot, but it would one more thing to configure when working with Telefunc & TypeScript. And it does not seem to have that much ongoing development or maintenence, it might stop working with a future version of TypeScript. 

Here are some ideas:

- The home grown approach: use the Compiler API at runtime during `shieldApply`: read inline parameters types in `*.telefunc.ts` files, parse AST and make appropiate call to `shield`. We'd be building  rather limited version of [tst-reflect](https://github.com/Hookyns/tst-reflect) and [typescript-rtti](https://github.com/typescript-rtti/typescript-rtti).

- The middle ground: ship a script that generates a file from all `*.telefunc.ts` files that we can use at runtime to check the type parameters. We could use [tst-reflect](https://github.com/Hookyns/tst-reflect) with the Compiler API.

- Embrace `ttypescript` fully, emit runtime type info for all types in the project. Upside: you can type and share complex type parameters more easily across telefunctions.

**Comment by louwers** [comment]
Another interesting option [ts-morph](https://ts-morph.com/details/types). A library that makes a the public Typescript API a lot easier to use. Seems very well mantiained and a lot more popular than the libraries mentioned above. Also, since Telefunc only will deal with serializable types they are probably overkill.

I think I'll hack something together using this library.

**Comment by brillout** [comment]
We already transform the `.telefunc.js` files to the client-side.

Depending on the Stack, we could as well transform `.telefunc.js` files for the server-side. We can, for example, do this with Vite.

This won't work for some stacks though (e.g. React Native).

The convenient thing though is that we won't need ttypescript then.

I'm wondering: how much simpler would it be if we only support inline types?

```ts
// *.telefunc.js

// Supported
export async function hello(name: string) {
}

// Not supported
type Name = string
export async function hello(name: Name) {
}
```

I'm being pushy about inline types because I think it's a viable solution for the (vast) majority of telefunctions.

We can go either way, but I think we should at least keep ourselves open to inline types only.

**Comment by louwers** [comment]
Here's the answer to that :wink: 

```ts
// hello.telefunc.ts

export async function hello(name: string) {
}

type Name = string
export async function hello2(name: Name) {
}
```

```ts
import { Project } from "ts-morph"

const project = new Project()
project.addSourceFileAtPath("hello-world.ts")

const sourceFile = project.getSourceFile("hello-world.ts")!

const parameters = sourceFile.getFunctions().filter(fun => fun.isExported).map(fun => {
  return fun.getParameters()
})

console.log(parameters[0][0].getType() === parameters[1][0].getType())
```

**Comment by louwers** [comment]
https://github.com/runem/ts-simple-type :eyes: 

I use `ts-morph` to load the telefunc parameter types in memory. I use `ts-simple-type` to do the type check at runtime. :sunglasses: 

```ts
// hello.telefunc.ts
type MyParam = {
  name: string
  age: number
  isCool: boolean
}

export async function hello(name: MyParam) {}

export async function hello2(name: {
  name: string
  age: number
  isCool: boolean
}) {}
```

```
import { Project } from "ts-morph"
import { isAssignableToValue } from "ts-simple-type";

const project = new Project()
project.addSourceFileAtPath("hello.telefunc.ts")

const sourceFile = project.getSourceFile("hello.telefunc.ts")!

const typeChecker = project.getTypeChecker().compilerObject

const parameters = sourceFile.getFunctions().filter(fun => fun.isExported).forEach(fun => {
  const val = {
    name: "Hello",
    age: 123,
    isCool: true
  }

  const val2 = {
    name: "Hello",
    age: "string",
    isCool: false
  }

  const type = fun.getParameters()[0].getType().compilerType

  console.log(isAssignableToValue(type, val, typeChecker))
  console.log(isAssignableToValue(type, val2, typeChecker))
})

// console.log(parameters[1][0].getType())

type MyParam = {
  name: string
  age: number
  isCool: boolean
}
```

**Comment by brillout** [comment]
That looks quite promising indeed.

Looks like we can soon ship a beta and already start playing with it 😍.

Actually about ttypescript: would it work for a Vite stack? The thing is that Vite uses esbuild to transpile TS, so I don't think ttypescript can do its magic then?

What we can do is to add a server-side transfomer over there:

https://github.com/vikejs/telefunc/blob/4aa22608b5be742a6b1c5ff78c8535a41aa32547/telefunc/node/vite/transform.ts#L17-L19

In Vite's world `ssr` means server-side code.

We'll probably need to split the transformer plugin into two plugins: one with `enforce: 'pre'` for server-side code (so that we can access the TS types) and one without `enforce` for client-side code. (The client-side transfomer uses `es-module-lexer` which will probably choke it receives typescript source code, e.g. `import { foo, type Foo } from './foo'`.)

https://vitejs.dev/guide/api-plugin.html#plugin-ordering.

**Comment by louwers** [comment]
No ttypescript needed for the approach above. :-)

**Comment by brillout** [comment]
👍 

In Vite's world, the `.ts` files are transpiled by Vite. This means we merely have to implement a Vite transformer and it will just work.

So for Vite stacks we don't need ttypescript, but I was thinking about other stacks such as React Native.

For React Native, we do transform `.telefunc.ts` files but only for the client-side. For the server-side we have to find other means to transform `.telefunc.ts` files.

Anyways, I'd suggest we focus on supporting only Vite for now.

**Comment by brillout** [comment]
Note that `vite-plugin-ssr` has an hydrid approach: some server files are processed by Vite, some not. E.g. code for Express.js is not processed by Vite, while server-side data fetching hooks are processed by Vite.

**Comment by louwers** [comment]
What if we skip these wonky libraries use Typescript itself for type checking. :smile: 


```ts
// hello.telefunc.ts
export async function doSomething(name:  {hello: string, world: true}, arg2: string) {

}
```

```ts
import { Project } from "ts-morph"

function toConstExpression(arg: unknown): string {
  return `${JSON.stringify(arg)} as const`
}

/**
 * Type check arguments of unknown type passed to a telefunc
 * @param telefuncFile File where telefunc is defined
 * @param telefuncName Name of the telefunc
 * @param args arguments passed to telefunc
 * @returns an array of errors, empty array when passed arguments have correct type
 */
function typeCheckTelefuncArgs(telefuncFile: string, telefuncName: string, ...args: unknown[]): string[] {
  const project = new Project({
    compilerOptions: {
      strict: true
    }
  })
  
  const sourceFile = project.addSourceFileAtPath(telefuncFile)
  if (!sourceFile) throw new Error(`Source file '${telefuncFile}' not found`)

  sourceFile.addVariableStatement({
    declarations: [{
      name: 'args',
      initializer: `[${args.map(arg => (toConstExpression(arg))).join(", ")}]`,
      trailingTrivia: " as const"
    }]
  })
  sourceFile.addStatements(`${telefuncName}(...args)`)
  return sourceFile.getPreEmitDiagnostics().map(diagnostic => {
    const messageText = diagnostic.getMessageText()
    return typeof messageText === 'string' ? messageText : messageText.getMessageText()
  })
}
```
Say you call it like this:

```ts
const errors = typeCheckTelefuncArgs("hello.telefunc.ts", "doSomething",
  { 'hello': '123', 'world': true },
  'second-arg')
```

Then it will generate a source file in memory like this:

```
export async function doSomething(name:  {hello: string, world: true}, arg2: string) {

}

let args = [{"hello":"123","world":true} as const, "second-arg" as const] as const;
doSomething(...args)
```

Then it will type check it, and return the errors (if any).

**Comment by louwers** [comment]
To avoid having to serialize when we already have (S-)JSON I can use a little helper:

```ts
type SJsonString<S extends string> = S extends '!undefined'
  ? undefined
  : (S extends `!Date:${string}`
    ? Date
    : (S extends `!RegExp:${string}`
      ? RegExp
      : S))

type A = SJsonString<'!undefined'>  // undefined

type B = SJsonString<'!Date:2022-04-16T18:06:40.450Z'>  // Date

type C = SJsonString<'!RegExp:/test/'>  // RegExp

type D = SJsonString<'OtherString'> // "OtherString"

type SJsonify<T> = {
    [K in keyof T]: T[K] extends string ? SJsonString<T[K]> : SJsonify<T[K]>
}

const data = {
    someDate: "!Date:2022-04-16T18:06:40.450Z",
    nested: {
        someBool: true,
        someUndef: "!undefined"
    }
} as const

type Z = SJsonify<typeof data>  // { someDate: Date, nested: { someBool: true, someUndef: undefined}}
```

:sunglasses:

**Comment by brillout** [comment]
That's quite a clever trick. I like it. The main question is whether it's performant enough? TypeScript can be slow if the types are complex.

This will most likely work if only support inline types though.

> Should I do some benchmarks to see if this is performant enough?

👍

Note that it's not about average performance but worst case performance.

It needs to be reliable: we can't use a threshold for the rare cases type checking takes too long. We have to ensure the performance is always acceptable, no matter how complex the type.

Maybe restricting to inline types is the answer? Maybe it's just performant enough, even for complex types? Maybe having some heuristic to determine how complex a type is and throw an error at compile-time if it's too complex?

>  project.addSourceFileAtPath(telefuncFile)

This means we don't support types defined outside of the `*.telefunc.ts`, right? Seems fine to me. (Because telefunctions have, for the vast majority of cases, simple parameters.)

As for serialization, we don't have to handle it: the arguments passed to `typeCheckTelefuncArgs()` are already deserialized. The question is more whether `toConstExpression()` supports any JSON-S data?

**Comment by louwers** [comment]
> The question is more whether toConstExpression() supports any JSON-S data?

The idea was to dump JSON-S serialized data in a Typescript source file with `as const` to get the type (mapping over it, so `Date` and `RegExp` would also be there).

But... Getting about 400ms with the example above. :snail:

**Comment by brillout** [comment]
That's quite slow indeed.

I guess we're back to square 1.

Using the TypeScript compiler alone, we can still get inline types working though, right?

```ts
// Not supported
// (Note that this is not idiomatic Telefunc usage. It's considered bad practice.)
function updateTodo(args: TodoProps) {
}

// Supported
// (This is idiomatic Telefunc usage. Event based telefunctions.)
function onTodoTextChange(text: string) {
}

// Supported
// (This is idiomatic and will soon been the standard way of doing data fetching with vite-plugin-ssr 0.4.)
function fetchTodoList({ isCompleted, paginationDate }: { isCompleted: boolean, paginationDate: Date }) {
}
```

**Comment by louwers** [comment]
Here is another crazy idea: use Typescript to generate the code that calls `shield`.

https://tsplay.dev/weX8XW

**Comment by brillout** [comment]
This is indeed an exotic idea :-) But could work... So the plan is to use `ts-morph` to evaluate and get that string type? I wonder: could we use `ts-morph` to evaluate the type of the arguments to do the same but without generating that string type?

> Output it to (e.g.) `hello.shield.telefunc.ts`, and then you'd have to put an import on top of every `*.telefunc.ts`:

Or we could use a transformer. That should let us achieve the same result but without having to generate a new file.

**Comment by bartenra** [comment]
> So the plan is to use `ts-morph` to evaluate and get that string type?

Exactly.

> I wonder: could we use ts-morph to evaluate the type of the arguments to do the same but without generating that string type?

Well. You can get the type of the arguments no problem! The difficulty is parsing the compiler representation of a type to a representation that allows you to work with it. In my very limited experience so far, analyzing types with Typescript is much easier and much more reliable, than using the (public) compiler API. [Here](https://github.com/runem/ts-simple-type/blob/master/src/transform/to-simple-type.ts) you can see a brave effort. That library has 35000 tests but I was able to find problematic types quite easily.

**Comment by brillout** [comment]
I see.

**Comment by louwers** [comment]
It's looking good though :smile: 

```ts
type cases = [
  Expect<Equal<ShieldStr<string>, 't.string'>>,
  Expect<Equal<ShieldStr<number>, 't.number'>>,
  Expect<Equal<ShieldStr<boolean>, 't.boolean'>>,
  Expect<Equal<ShieldStr<Date>, 't.date'>>,
  Expect<Equal<ShieldStr<any>, 't.any'>>,
  Expect<Equal<ShieldStr<string | null>, 't.nullable(t.string)'>>,
  Expect<Equal<ShieldStr<string | undefined>, 't.optional(t.string)'>>,
  Expect<Equal<ShieldStr<[number, boolean]>, 't.tuple(t.number, t.boolean)'>>,
  Expect<Equal<ShieldStr<number[]>, "t.array(t.number)">>,
  Expect<Equal<ShieldStr<{age: number, hasBike: boolean}>, '{ hasBike: t.boolean, age: t.number }'>>,
  Expect<Equal<ShieldStr<{age: number, hasBike?: boolean}>, '{ hasBike: t.optional(t.boolean), age: t.number }'>>,
  Expect<Equal<ShieldStr<'test'>, "t.const('test')">>,
  Expect<Equal<ShieldStr<true>, "t.const(true)">>,
  Expect<Equal<ShieldStr<false>, "t.const(false)">>,
  Expect<Equal<ShieldStr<'test' | 'hi'>, "t.union(t.const('hi'), t.const('test'))">>,
  Expect<Equal<ShieldStr<number | string>, "t.union(t.number, t.string)">>,
  Expect<Equal<ShieldStr<{isGood: boolean} | string>, "t.union({ isGood: t.boolean }, t.string)">>,
]
```
I don't think this would be possible without Tail-Recursion Elimination on Conditional Types introduced in Typescript 4.5.

**Comment by brillout** [comment]
Quite satisfying to see these unit tests increasing 😀.

This is all increasingly promising 👀.

**Comment by louwers** [comment]
I got all the `shield()` types now. :+1: https://tsplay.dev/Wy4xxm

I can move forward with intergrating it now.

**Comment by brillout** [comment]
💯

For integration I'd suggest we re-use the transformer architecture. For Vite:

> What we can do is to add a server-side transfomer over there:
> 
> https://github.com/vikejs/telefunc/blob/4aa22608b5be742a6b1c5ff78c8535a41aa32547/telefunc/node/vite/transform.ts#L17-L19

This means that Vite users don't have to do any, it just works.

For stacks that don't use a server-side bundler, maybe we can use ttypescript?

**Comment by brillout** [comment]
Btw. a potential solution for the problem of the emit info being truncated:

I can see that the threshold for truncating to be some `const` variable in TypeScript.

If that's the case then it's easy to change it to some big number `9999` in a post-install script. (I believe TypeScript ships as JavaScript and not binary; should be easy to patch TypeScript.)

We pin TypeScript and check if the patch still works when we use a newer TypeScript version.

**Comment by brillout** [comment]
Implemented by https://github.com/vikejs/telefunc/pull/25. Released in `0.1.20`.
